src/algebra/square.lean

```diff
@@ -0,0 +1,201 @@
+/-
+Copyright (c) 2022 Kevin H. Wilson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin H. Wilson
+-/
+import ring_theory.int.basic
+
+/-!
+# Square elements of monoids
+An element `x` of a monoid is a square when there is another element `y` such that `y * y = x`.
+
+## Main Definitions
+ - `square r` indicates that `∃ s, s * s = r`.
+
+## Main Results
+ - `nat.square_iff_factorization_even`: `x` is a `square` iff every entry in its factorization
+   is even
+ - `nat.square_mul`: Being `square` is multiplicative
+
+## TODO
+
+There are many more facts about squares that generalize to arbitrary `unique_factorization_monoid`s.
```

Comment:
About half of the file deals specifically deals with naturals. I'm not very familiar with UFDs, but I'd assume many of these results can be generalized. Are these included in this TODO remark?

Reply:
To this end, I've added a _long_ comment. Please tell me if it's too much!